crates/interledger-service-util/src/echo.rs

```
@@ -111,21 +113,23 @@ where
                 return Box::new(err(RejectBuilder {
                     code: ErrorCode::F01_INVALID_PACKET,
                     message: b"Could not read source address.",
-                    triggered_by: self.ilp_address.borrow(),
+                    triggered_by: Some(&self.ilp_address),
                     data: &[],
                 }
                 .build()));
             }
         };
 
+        let source_address = Address::try_from(source_address).unwrap();
```

Is it safe to `unwrap` here? If the `source_address: &[u8]` is an invalid address this will panic.

---

Yeah I was going to fix it, https://github.com/emschwartz/interledger-rs/pull/75/commits/e56d53d3477571f78a5a3d9977621f7c954764fd is my initial approach, following how the rest of the file handles errors, but I do not really like it. Do you perhaps have any suggestions for that

---

The first check actually just reads it from the buffer, but the actual checking that it conforms to the spec is not done until the second try_from, so I do not think it's unnecessary

---

Ah, sorry. Misread the match statement. If there's an error here it should return an F01 error just like the above check. Instead of using the matches

---

```rust
        let source_address = match reader.read_var_octet_string() {
            Ok(value) => Address::try_from(value).map_err(|_| {
                Box::new(err(RejectBuilder {
                    code: ErrorCode::F01_INVALID_PACKET,
                    message: b"Could not parse source address from Echo packet",
                    triggered_by: Some(&self.ilp_address),
                    data: &[],
                }
                .build()))
            })?,
            Err(error) => {
                eprintln!("Could not read source address: {:?}", error);
                return Box::new(err(RejectBuilder {
                    code: ErrorCode::F01_INVALID_PACKET,
                    message: b"Could not read source address.",
                    triggered_by: Some(&self.ilp_address),
                    data: &[],
                }
                .build()));
            }
        };
```

This was what I had done initially, but the question mark operator doesn't behave well with futures it seems

```
rror[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
   --> crates/interledger-service-util/src/echo.rs:110:26
    |
110 |               Ok(value) => Address::try_from(value).map_err(|_| {
    |  __________________________^
111 | |                 Box::new(err(RejectBuilder {
112 | |                     code: ErrorCode::F01_INVALID_PACKET,
113 | |                     message: b"Could not parse source address from Echo packet",
...   |
117 | |                 .build()))
118 | |             })?,
    | |_______________^ cannot use the `?` operator in a function that returns `std::boxed::Box<dyn futures::future::Future<Error = interledger_packet::packet::Reject, Item = interledger_packet::packet::Fulfill> + std::marker::Send>`
    |
    = help: the trait `std::ops::Try` is not implemented for `std::boxed::Box<dyn futures::future::Future<Error = interledger_packet::packet::Reject, Item = interledger_packet::packet::Fulfill> + std::marker::Send>`
    = note: required by `std::ops::Try::from_error`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: Could not compile `interledger-service-util`.
```

---

Right. Ahh the pleasure of working with Future types. Looking forward to async await